### ISSUE TITLE
`tuple(tuple)` loses size information

### ISSUE THREAD

**Issue by bersbersbers** [microsoft/pyright#6035]
**Describe the bug**
In the code example below, `return values` works fine, while `return tuple(values)` seems to ignore the information that `values` has two elements, just as required by the return type. So I am getting a size mismatch:
```
bug.py
  bug.py:3:12 - error: Expression of type "tuple[Literal[0, 1], ...]" cannot be assigned to return type "tuple[int, int]"
    "tuple[Literal[0, 1], ...]" is incompatible with "tuple[int, int]"
      Tuple size mismatch; expected 2 but received indeterminate (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

**Code or Screenshots**
If possible, provide a minimal, self-contained code sample (surrounded by triple back ticks) to demonstrate the issue. The code should define or import all referenced symbols.

```python
def return_tuple() -> tuple[int, int]:
    values = (0, 1)
    return tuple(values)
```

If your code relies on symbols that are imported from a third-party library, include the associated import statements and specify which versions of those libraries you have installed.

**VS Code extension or command-line**
`pyright 1.1.329` from command line

**Comment by bersbersbers** [comment]
Even this simplified version fails:

```python
def return_tuple() -> tuple[int, int]:
    return tuple((0, 1))
```

**Comment by erictraut** [comment]
Yes, this is working as intended. If you use a call to the `tuple` constructor, pyright uses the type information provided by the builtins.pyi type stub. This definition obscures the length of the tuple. 

If you want to retain the length, you should use a tuple expression (e.g. `(0, 1)`). Type checkers have special-case logic to handle tuple expressions, and this logic retains tuple length information.
